File: ui/src/main/java/com/peakmain/ui/imageLoader/glide/GlideLoader.java
package com.peakmain.ui.imageLoader.glide;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.net.Uri;
import android.os.Build;
import android.text.TextUtils;
import android.util.Log;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.resource.bitmap.RoundedCorners;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.SimpleTarget;
import com.peakmain.ui.imageLoader.ILoader;
import com.peakmain.ui.utils.LogUtils;

import java.io.File;

/**
 * author ：Peakmain
 * createTime：2020/2/22
 * mail:2726449200@qq.com
 * describe：Glide的
 */
public class GlideLoader implements ILoader {

    /**
     * 返回一个请求的配置
     */
    private RequestOptions getRequestOptions(int resId) {

        return getRequestOptions(resId, false);
    }

    @SuppressLint("CheckResult")
    private RequestOptions getRequestOptions(int resId, boolean isSkipCache) {
        RequestOptions options = new RequestOptions();
        if (isSkipCache) {
            options.skipMemoryCache(true);
            options.diskCacheStrategy(DiskCacheStrategy.NONE);
        }
        //设置一张占位图
        options.placeholder(resId);
        return options;
    }


    @Override
    public void displayImage(Context context, String url, ImageView view, int desId) {
        RequestOptions options = getRequestOptions(desId);
        loadImage(context, url, view, options);
    }

    @Override
    public void displayImage(Context context, String url, ImageView view, int desId, boolean isSkipCache) {
        RequestOptions options = getRequestOptions(desId, isSkipCache);
        loadImage(context, url, view, options);
    }

    @Override
    public void displayImageRound(Context context, String url, ImageView view, int corner, int desId) {
        RequestOptions options = getRequestOptions(desId).transform(new RoundedCorners(corner));
        loadImage(context, url, view, options);
    }

    @Override
    public void displayImage(Context context, Uri url, ImageView view, int desId) {
        RequestOptions options = getRequestOptions(desId);
        loadImage(context, url, view, options);
    }

    @SuppressLint("CheckResult")
    @Override
    public void displayImage(Context context, String url, ImageView view, int height, int width, int desId) {
        RequestOptions options = getRequestOptions(desId);
        options.override(width, height);
        loadImage(context, url, view, options);
    }

    @SuppressLint("CheckResult")
    @Override
    public void displayImage(Context context, String url, ImageView view, int height, int width, float sizeMultiplier, int desId) {
        RequestOptions options = getRequestOptions(desId);
        options.sizeMultiplier(sizeMultiplier);
        options.override(width, height);
        loadImage(context, url, view, options);
    }

    @Override
    public void displayLocalImage(Context context, String url, ImageView view, int desId) {
        if (TextUtils.isEmpty(url)) {
            return;
        }
        RequestOptions options = getRequestOptions(desId);
        File file = new File(url);
        if (file.exists()) {
            loadImage(context, file, view, options);
        }
    }

    @Override
    public void displayImage(Context context, Uri url, SimpleTarget simpleTarget) {
        try {
            if (context != null) {
                if (context instanceof Activity) {
                    Activity activity = (Activity) context;
                    if (!isDestroy(activity)) {
                        Glide.with(activity).load(url).into(simpleTarget);
                    }
                } else {
                    Glide.with(context).load(url).into(simpleTarget);
                }
            }
        } catch (Exception e) {
            LogUtils.d("加载图片出错:" + e.getMessage());
        }
    }

    @Override
    public void clearMemory(Context context) {
        Glide.get(context).clearMemory();
    }

    @Override
    public void clearDiskCache(Context context) {
        Glide.get(context).clearDiskCache();
    }

    /**
     * 程序在内存清理的时候执行
     */

    @Override
    public void trimMemory(Context context, int level) {
        Glide.get(context).trimMemory(level);
    }

    @Override
    public void clearAllMemoryCaches(Context context) {
        Glide.get(context).onLowMemory();
    }

    /**
     * 恢复请求
     */
    @Override
    public void resumeRequest(Context context) {
        Glide.with(context).resumeRequests();
    }

    @Override
    public void pauseRequest(Context context) {
        Glide.with(context).pauseRequests();
    }

    private void loadImage(Context context, Object url, ImageView view, RequestOptions options) {
        try {
            if (context != null) {
                if (context instanceof Activity) {
                    Activity activity = (Activity) context;
                    if (!isDestroy(activity)) {
                        Glide.with(activity).load(url).apply(options).into(view);
                    }
                } else {
                    Glide.with(context).load(url).apply(options).into(view);
                }
            }
        } catch (Exception e) {
            LogUtils.d("加载图片出错：" + e.getMessage());
        }
    }

    //判断activity有没有被销毁
    private boolean isDestroy(Activity activity) {
        if (activity == null || activity.isFinishing() ||
                (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1 && activity.isDestroyed())) {
            return true;
        } else {
            return false;
        }
    }
}

File: ui/src/main/java/com/peakmain/ui/imageLoader/glide/GlideLoader.kt
package com.peakmain.ui.imageLoader.glide

import android.annotation.SuppressLint
import android.content.Context
import android.net.Uri
import android.text.TextUtils
import android.widget.ImageView
import androidx.annotation.DrawableRes
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.model.GlideUrl
import com.bumptech.glide.load.model.LazyHeaders
import com.bumptech.glide.load.resource.bitmap.RoundedCorners
import com.bumptech.glide.request.RequestOptions
import com.peakmain.ui.imageLoader.ILoader
import com.peakmain.ui.imageLoader.transform.CornerTransform
import com.peakmain.ui.utils.LogUtils
import java.io.File
import java.net.MalformedURLException
import java.net.URL

/**
 * author ：Peakmain
 * createTime：2020/2/22
 * mail:2726449200@qq.com
 * describe：Glide的
 */
class GlideLoader : ILoader {
    private var userAgent: String? = null

    /**
     * 返回一个请求的配置
     */
    private fun getRequestOptions(@DrawableRes resId: Int): RequestOptions {
        return getRequestOptions(resId, false)
    }

    @SuppressLint("CheckResult")
    private fun getRequestOptions(@DrawableRes resId: Int, isSkipCache: Boolean): RequestOptions {
        val options = RequestOptions()
        if (isSkipCache) {
            options.skipMemoryCache(true)
            options.diskCacheStrategy(DiskCacheStrategy.NONE)
        }
        //设置一张占位图
        options.placeholder(resId)
        return options
    }

    override fun userAgent(userAgent: String?) {
        this.userAgent = userAgent
    }

    override fun displayImage(
        context: Context?,
        url: String?,
        view: ImageView?,
        @DrawableRes desId: Int
    ) {
        val options = getRequestOptions(desId)
        loadImage(context, url, view, options)
    }

    override fun displayImage(
        context: Context?,
        url: String?,
        view: ImageView?,
        @DrawableRes desId: Int,
        isSkipCache: Boolean
    ) {
        val options = getRequestOptions(desId, isSkipCache)
        loadImage(context, url, view, options)
    }

    override fun displayImageRound(
        context: Context?,
        url: String?,
        view: ImageView?,
        corner: Int,
        @DrawableRes desId: Int
    ) {
        val options = getRequestOptions(desId).transform(RoundedCorners(corner))
        loadImage(context, url, view, options)
    }


    override fun displayImage(
        context: Context?,
        url: Uri?,
        view: ImageView?,
        @DrawableRes desId: Int
    ) {
        val options = getRequestOptions(desId)
        loadImage(context, url, view, options)
    }

    @SuppressLint("CheckResult")
    override fun displayImage(
        context: Context?,
        url: String?,
        view: ImageView?,
        height: Int,
        width: Int,
        @DrawableRes desId: Int
    ) {
        val options = getRequestOptions(desId)
        options.override(width, height)
        loadImage(context, url, view, options)
    }

    @SuppressLint("CheckResult")
    override fun displayImage(
        context: Context?,
        url: String?,
        view: ImageView?,
        height: Int,
        width: Int,
        sizeMultiplier: Float,
        @DrawableRes desId: Int
    ) {
        val options = getRequestOptions(desId)
        options.sizeMultiplier(sizeMultiplier)
        options.override(width, height)
        loadImage(context, url, view, options)
    }

    override fun displayImage(
        context: Context?,
        url: String?,
        imageView: ImageView?,
        requestOptions: RequestOptions
    ) {
        if (context == null || imageView == null || TextUtils.isEmpty(url)) return
        displayImage(url!!, imageView, requestOptions)
    }


    override fun displayLocalImage(
        context: Context?,
        url: String?,
        view: ImageView?,
        @DrawableRes desId: Int
    ) {
        if (TextUtils.isEmpty(url)) {
            return
        }
        val options = getRequestOptions(desId)
        val file = File(url!!)
        if (file.exists()) {
            loadImage(context, file, view, options)
        }
    }

    private fun displayImage(
        url: String,
        imageView: ImageView,
        options: RequestOptions
    ) {
        var url1: URL? = null
        try {
            url1 = URL(url)
        } catch (e: MalformedURLException) {
            e.printStackTrace()
        }
        url1?.let {
            val glideUrl =
                GlideUrl(it, LazyHeaders.Builder().addHeader("User-Agent", userAgent!!).build())
            Glide.with(imageView.context.applicationContext)
                .asDrawable()
                .load(glideUrl)
                .apply(options)
                .into(imageView)
        }
    }

    override fun clearMemory(context: Context?) {
        if (context == null) return
        Glide.get(context).clearMemory()
    }

    override fun clearDiskCache(context: Context?) {
        if (context == null) return
        Glide.get(context).clearDiskCache()
    }

    /**
     * 程序在内存清理的时候执行
     */
    override fun trimMemory(context: Context?, level: Int) {
        context?.let {
            Glide.get(it).trimMemory(level)
        }
    }

    override fun clearAllMemoryCaches(context: Context?) {
        context?.let {
            Glide.get(it).onLowMemory()
        }
    }

    /**
     * 恢复请求
     */
    override fun resumeRequest(context: Context?) {
        context?.let {
            Glide.with(it).resumeRequests()
        }
    }

    override fun pauseRequest(context: Context?) {
        context?.let {
            Glide.with(it).pauseRequests()
        }
    }

    /**
     * 设置图片的四周圆角
     */
    override fun displayImageRound(
        context: Context?,
        url: String?,
        view: ImageView?,
        @DrawableRes desId: Int,
        leftTop: Boolean,
        rightTop: Boolean,
        leftBottom: Boolean,
        rightBottom: Boolean,
        roundRadius: Float
    ) {
        if (context == null || TextUtils.isEmpty(url) || view == null) return
        val cornerTransform = CornerTransform(context, roundRadius)
        cornerTransform.setCorner(leftTop, rightTop, leftBottom, rightBottom)
        val options = RequestOptions().placeholder(desId)
            .transform(cornerTransform)
        displayImage(url!!, view, options)
    }

    private fun loadImage(context: Context?, url: Any?, view: ImageView?, options: RequestOptions) {
        if (context == null || view == null) return
        try {
            if (url is String && !TextUtils.isEmpty(userAgent)) {
                displayImage(url, view, options)
            } else {
                Glide.with(context).load(url).apply(options).into(view)
            }
        } catch (e: Exception) {
            LogUtils.e("加载图片出错：" + e.message)
        }
    }

}